### PR TITLE
[WIP] Enable projecting persistent buffers to inputs to save memory.

### PR DESCRIPTION
**Pull request by csarofeen**
Mainly for batch norm backwards, but generalized in case we hit this anywhere else. If the persistent buffers are smaller if we project them to the inputs (if we can), especially in fp16, the buffers are projected to the inputs.

Only thing left on this PR is to write some tests for the persistent buffer calculations, and to enable padding to next warp.

Batch Norm BWD sees the biggest perf gains from this PR as expected (with padding to warp):
image

The regressions are unfortunate but they don't seem unreasonable to me. These are cases where we were leaking because our buffer calculation was incorrect but we were running persistently. So we were relying on local memory to do the right thing, and the compiler to not run out of registers with the sizes we ran with.

Overall perf change with padding to warp:
image

### CODE REVIEW

**Comment by csarofeen** [comment]
This was merged in https://github.com/csarofeen/pytorch/pull/1237
